### ISSUE TITLE
Improve how connections are defined in settings

### ISSUE THREAD

**Issue by gjsjohnmurray** [intersystems-community/vscode-objectscript#135]
I am opening this issue to start a discussion.

## Current situation

An extension's settings can be held in several locations, such as:

* `settings.json` for the user
* `XYZ.code-workspace` file if used
* `.vscode/settings.json` in the current root folder

More details in VS Code docs [here](https://code.visualstudio.com/docs/getstarted/settings).

In this extension the settings for how to connect to an InterSystems server are currently (as at 0.8.6) held in a structure that can only define a single connection. Below is the JSON of the default instance of this structure:

```json
"objectscript.conn": {
    "active": false,
    "host": "localhost",
    "port": 52773,
    "username": "_SYSTEM",
    "password": "SYS",
    "ns": "USER",
    "https": false,
    "docker-compose": {
        "service": "",
        "internalPort": 52773,
        "file": "docker-compose.yml"
    },
    "links": {}
}
```

This is a significant restriction at user level. On my workstation I can only define a single connection.

When the connection is defined per workspace (in XYZ.code-workspace or in .vscode/settings.json) the limitation is tolerable, but it obstructs the use of [multi-root workspaces](https://code.visualstudio.com/docs/editor/multi-root-workspaces) and the [isfs scheme](https://github.com/intersystems-community/vscode-objectscript/wiki/Server-side-editing) to span multiple servers.

Contrast this with how [InterSystems IRIS Server Manager](https://docs.intersystems.com/irislatest/csp/docbook/Doc.View.cls?KEY=GSA_connect_remote#GSA_connect_remote_define) allows definition of multiple named connections.

## Proposals

Below is intended as a discussion-starter.

1. Use an array, subscripted by a user-defined name for each connection. The extension's definition of the settings it contributes can validate names against a pattern, e.g.

```json
"intersystems.servers": {
    "type": "object",
    "description": "Target environments. Each property names a server and holds nested properties defining how to connect to it.\nServer names may only contain characters 'A' to 'Z', 'a' to 'z', digits, '.', '-' and '_' characters.",
    "scope": "resource",
    "default": {
        "localhost": {
            "host": "localhost"
        }
    },
    "patternProperties": {
        "^[A-Za-z0-9_\\.\\-]+$": {
            "type": "object",
            "description": "A server definition, with properties that specify how to connect to it",
            "properties": {
                "host": {
                    "type": "string",
                    "description": "Hostname or IP address of target InterSystems environment.",
                    "anyOf": [
                        {
                            "format": "hostname"
                        },
                        {
                            "format": "ipv4"
                        },
                        {
                            "format": "ipv6"
                        }
                    ],
                    "default": "localhost"
                },

                ...

            }
        }
    },
    "additionalProperties": false
},

```

2. Align connection properties with the settings that the existing Server Manager lets users define. For example, these cater for:
    * Connecting to InterSystems superserver as well as to to web server
    * InterSystems servers located behind a shared web server (i.e. not using the private Apache)

3. Separate the connection definition (which could be used by multiple extensions) from other settings (e.g. "links", and maybe also "docker-compose") that are extension-specific. Extensions should store connection-specific settings in their own array keyed by server name as defined in the `intersystems.servers` structure above.

**Comment by mvhenderson** [comment]
@gjsjohnmurray I totally agree - see #119 for my take on it which is primarily concerned with moving the details out of the workspace/folder settings in into user settings. This way we could checkin the .code-workspace file with project rules without exposing host/ports/credentials.

**Comment by MaxAbrahams** [comment]
I disagree - I think the current one-connection-per-workspace is correctly aligned with VSCODE's philosphy on what an IDE is supposed to do. I have found the use of multiple workspaces freeing compared to studio, atelier, etc. when it comes to managing both my local filesystem and the database's compiled classes.

I think this proposed solution is significantly more complicated than a small change in the user's organization strategy and development mindset. Im not sure I understand your point about multi-root workspaces being needed over using multiple workspaces.

**Comment by gjsjohnmurray** [comment]
@MaxAbrahams suppose you have two IRIS instances, DEV and TEST, each with an APP namespace on it. With the current `objectscript.conn` syntax I don't think you can create a single workspace that accesses both, like this (done with [Serenji](https://marketplace.visualstudio.com/items?itemName=georgejames.vscode-serenji))

![image](https://user-images.githubusercontent.com/6726799/82673216-66ccfb00-9c39-11ea-8fd1-f408c5a04246.png)

**Comment by bourette** [comment]
Hi John,

You should be able to create a settings.json in the folder for each
server/namespace.

**Comment by gjsjohnmurray** [comment]
@bourette I am specifically referring to the use of the [isfs scheme](https://github.com/intersystems-community/vscode-objectscript/wiki/Server-side-editing) to connect VS Code's Explorer directly to namespaces and operate server-side (think Studio, not Atelier).

**Comment by bourette** [comment]
Yes, for each isfs scheme I also have a work folder defined which contains the settings file.

James

**Comment by gjsjohnmurray** [comment]
> Yes, for each isfs scheme I also have a work folder defined which contains the settings file.

@bourette I don't understand what you mean. Please post the (sanitized) contents of a .code-workspace file that you open in VS Code to give you a multiroot workspace accessing multiple servers with the isfs scheme.

**Comment by MaxAbrahams** [comment]
@gjsjohnmurray right, i get that you can't have two connections in a single workspace. my question was  - Why is 2 workspaces, (i.e. one for DEV[APP] and one for TEST[APP] in your example) not an acceptable, if not preferable, workflow?

**Comment by MaxAbrahams** [comment]
if the complaint is simply the clutter of having two workspaces open simultaneously, consider a plugin like [this](https://gitlab.com/tomsaunders-tools/vscode-workspace-explorer)

**Comment by gjsjohnmurray** [comment]
@MaxAbrahams my proposal will not prevent you from working in the way you find preferable. The opposite is not true.

**Comment by bourette** [comment]
@JohnMurray here is the sample you were looking for.  Basically it has two
namespaces on two different ports, they could easily be different systems
by defining the host as well.  It's a bit overkill because the workspace
defines the connection and then I also define the same port in Sys1.

**Comment by bspead** [comment]
@bourette  - where is the sample?  Should it be an attachment or inline? (I don't see either in my view of the issue)

**Comment by bourette** [comment]
Sorry here it is.
[VSSample.zip](https://github.com/intersystems-community/vscode-objectscript/files/4691125/VSSample.zip)

**Comment by gjsjohnmurray** [comment]
#120 may also be relevant to this discussion.

**Comment by evshvarov** [comment]
@gjsjohnmurray why not just open the second VSCode window?

**Comment by gjsjohnmurray** [comment]
> 
> 
> @gjsjohnmurray why not just open the second VSCode window?

For one example of something that can only be done within a single window, see https://github.com/intersystems-community/vscode-objectscript/issues/135#issuecomment-632694059

**Comment by MaxAbrahams** [comment]
you could simply drag the file you want to diff into a the second workspace window and diff them.
![image](https://user-images.githubusercontent.com/11665665/83418631-ef922680-a3f1-11ea-83f2-425bb4d88a93.png)

or any number of different ways of running  a diff.

**Comment by isc-jamien** [comment]
> @gjsjohnmurray why not just open the second VSCode window?

It's a common use-case to want to work in multiple namespaces, having them in the same VSCode workspace has a host of benefits.

One of the biggest complaints about Studio was that you had to open two copies to work with two different namespaces at the same time. When there are many namespaces in play then we  shouldn't have to have multiple instances of VSCode cluttering up the desktop.

**Comment by isc-jamien** [comment]
As regards the multiple named configurations. This is an essential feature and will ease migration of our Studio user base to VSCode because we can easily implement a means to import the pre-existing settings from the server manager registry settings on windows. This would be a real boon.

**Comment by bourette** [comment]
Couldn't agree more.

It's already possible to setup multiple connections using .vscode/settings.json in folders, why not just make it easier by defining the connection on the folder definition.

On Jun 1, 2020, at 10:31 AM, isc-jamien <notifications@github.com<mailto:notifications@github.com>> wrote:



As regards the multiple named configurations. This is an essential feature and will ease migration of our Studio user base to VSCode because we can easily implement a means to import the pre-existing settings from the server manager registry settings on windows. This would be a real boon.

**Comment by gjsjohnmurray** [comment]
To progress this issue a helper extension is being created that will be responsible for defining the structure of a settings object named `intersystems.servers`

The helper extension project is at https://github.com/intersystems-community/intersystems-servermanager

**Comment by gjsjohnmurray** [comment]
#172 implemented this and can be previewed in the beta vsix at https://github.com/intersystems-community/vscode-objectscript/releases/tag/v0.8.7-beta.22

**Comment by gjsjohnmurray** [comment]
Shipped in 0.8.7
